lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -545,84 +546,203 @@ def as_dict_with_promql_labels(self):
+            topology: juju topology information.
+            group_name: original group name to keep as part of the new augmented group name
+
+        Returns:
+            New group name, augmented by juju topology and relative path.
         """
-        relpath = os.path.relpath(os.path.dirname(path), str(dir_path))
+        rel_path = os.path.relpath(os.path.dirname(file_path), root_path)
+        rel_path = "" if rel_path == "." else rel_path.replace(os.path.sep, "_")
 
         # Generate group name:
-        #  - prefix, from the relative path of the rule file;
         #  - name, from juju topology
-        prefix = "" if relpath == "." else relpath.replace(os.path.sep, "_") + "_"
-        return "{}{}_alerts".format(prefix, topology.identifier)
+        #  - suffix, from the relative path of the rule file;
+        group_name_parts = [topology.identifier, rel_path, group_name, "alerts"]
+        return "_".join(filter(None, group_name_parts))
+
+    @staticmethod
+    def _from_dir(dir_path: str, topology: JujuTopology, recursive: bool) -> List[dict]:
+        """Read all rule files in a directory.
+
+        All rules from files for the same directory are loaded into a single
+        group. The generated name of this group includes juju topology.
+        By default, only the top directory is scanned; for nested scanning, pass `recursive=True`.
+
+        Args:
+            dir_path: directory containing *.rule files (alert rules without groups).
+            topology: juju topology information.
+            recursive: flag indicating whether to scan for rule files recursively.
+
+        Returns:
+            a list of dictionaries representing prometheus alert rule groups, each dictionary
+            representing an alert group (structure determined by `yaml.safe_load`).
+        """
+        alert_groups = []  # type: List[dict]
+
+        # Gather all alerts into a list of groups
+        paths = Path(dir_path).glob("**/*.rule" if recursive else "*.rule")
+        for file_path in filter(Path.is_file, paths):
+            alert_groups_from_file = AlertRules._from_file(dir_path, file_path, topology)
+            if alert_groups_from_file:
+                logger.debug("Reading alert rule from %s", file_path)
+                alert_groups.extend(alert_groups_from_file)
 
-    paths = Path(dir_path).glob("**/*.rule" if recursive else "*.rule")  # type: ignore
-    for path in filter(Path.is_file, paths):
-        if rule := load_alert_rule_from_file(path, topology):
-            logger.debug("Reading alert rule from %s", path)
-            alerts[_group_name(path)].append(rule)
+        return alert_groups if alert_groups else []
 
-    # Gather all alerts into a list of groups since Prometheus
-    # requires alerts be part of some group
-    groups = [{"name": k, "rules": v} for k, v in alerts.items()]
-    return groups
+    def add_from_path(self, path: str, *, recursive: bool = False):
```

Comment:
It is not clear why this method is required. All this method is doing is validating its argument and invoking `_from_dir()`. It seems odd to have two methods for reading all alert rules from a directory with the only difference being - One is and instance method that invoke the other which is a private static method. Why not have just one public instance method that reads all alert rules from the directory. Another possibility is to make the directory path a constructor argument to `AlertRules` object which then only needs to provide methods to return the alert rules in different formats.

Reply:
I made a small addition. Take a look again?